[TF:TRT] Enable TensorRT explicit precision (QDQ/QAT) support

christopherbate opened the pull request:
Adds TensorRT QDQ support ("explicit precision mode"), also sometimes referred to as "QAT support", referring to the training algorithm where QDQ nodes are used. From here on, we refer to the existing non-explicit precision pathway in the code base as the "dynamic range INT8" mode (DR INT8) and the new mode as the QDQ INT8 mode.

In the new mode, TF `QuantizeAndDequantize` operations are converted to TensorRT quantization scaling layers. Both the new QDQ mode logic as well as existing DR mode logic (`ConvertQuantize`) are moved into the file `convert/ops/quantization_ops.cc`. 

In addition, in QDQ mode it is necessary to prevent the existing Grappler optimizations invoked in `trt_convert.py` on the loaded SavedModel from folding frozen `QuantizeAndDequantizeV2` operations between weighted ops (Conv, Matmul ,etc) and the weight constants. Thus, we depend on the experimental Grappler rewriter config option `experimental_disable_folding_quantization_emulation` and will be affected if it is removed. The alternative is to allow Grappler folding of the QDQ and constant weights and inserting identity QDQ scale factors manually during TensorRT network construction, but the logic becomes extremely verbose .

A test suite is added in `convert/ops/quantization_ops_test.cc`. It builds a variety of sub-graph patterns and tests for conversion success. Because TRT QDQ mode has evolved significantly in terms of robustness and features between TRT 7 and TRT8, a set of test waive/skip policies are added indicating which patterns of use are appropriate for TRT7 vs TRT8

christopherbate commented:
Conflicts resolved

bixia1 commented on tensorflow/python/compiler/tensorrt/trt_convert.py:
```diff
@@ -167,11 +168,13 @@ def __new__(cls,
               use_calibration=True,
-              allow_build_at_runtime=True):
+              allow_build_at_runtime=True,
+              use_explicit_precision=False):
```
We will discuss the naming of this parameter at our meeting tomorrow.
Please add document for this new parameter to the document block above.
Also, it would be nice to split this change to a separated PR, and needed update to the compatibility tests related to the API.

christopherbate replied:
ok we also recently discussed not using this specific API, I'll remove API changes

bixia1 replied:
Per offline discussion, TF-TRT will decide whether to use use_explicit_precision based on TensorRT version.

bixia1 commented on tensorflow/compiler/tf2tensorrt/convert/ops/quantization_ops.cc:
```diff
@@ -0,0 +1,414 @@
+namespace convert {
+
+bool IsQuantizeAndDequantizeOp(const Node* node) {
+  // return TrtNodeValidator::quantize_ops->count(node->def().op()) != 0;
```
Remove this line.

bixia1 commented on tensorflow/compiler/tf2tensorrt/convert/ops/quantization_ops.h:
```diff
@@ -0,0 +1,76 @@
+  std::array<T, N> dequantize_scale;
+};
+
+// In TensorRT 7 and 8, only uniform tensor scaling is supported for
```
Shall we s/8/8.0/, because for > 8.0, per channel quantization is supported?

christopherbate replied:
For activations, only uniform scaling is supported.
For weights, per channel is supported for TRT >=8.0.

bixia1 commented on tensorflow/compiler/tf2tensorrt/convert/ops/quantization_ops.cc:
```diff
@@ -0,0 +1,414 @@
+}
+
+// Prepares the input for a QDQ node in explicit precision mode, returning a
+// ITensor pointer. If the input is weights, we convert to a ITensor by adding a
```
s/convert/convert it/

bixia1 commented on tensorflow/compiler/tf2tensorrt/convert/ops/quantization_ops.cc:
```diff
@@ -0,0 +1,414 @@
+    UniformQuantizationScales scales;
+  };
+
+  static Status ValidateExplicit(const std::vector<TRT_TensorOrWeights>& inputs,
```
It will be clearer to me what this function does if we name it "ValidataQDQForExplicitPrecision"

bixia1 commented:
I didn't review the change for the existing for to add use_explicit_precision, assuming that will be gone.

christopherbate commented:
Made a slight rebase error, will push up small diff as new commit

christopherbate commented on tensorflow/compiler/tf2tensorrt/convert/ops/quantization_ops_test.cc:
```diff
@@ -0,0 +1,542 @@
+Output AddOutput(Scope scope, Output input, int idx, bool add_qdq) {
+  Output out = input;
+  if (add_qdq) {
+    VLOG(0) << "Adding QDQ before output";
```
removed

christopherbate commented on tensorflow/compiler/tf2tensorrt/convert/ops/quantization_ops_test.cc:
```diff
@@ -0,0 +1,542 @@
+                          ops::BiasAdd::Attrs().DataFormat(data_format));
+  }
+
+  // We lower BN to sequence of {scale,shift} here.
```
Old comment, sorry. Removed

christopherbate commented on tensorflow/compiler/tf2tensorrt/convert/ops/quantization_ops_test.cc:
```diff
@@ -0,0 +1,542 @@
+        ops::Const(scope.WithOpName("bn_var"), 1.0f, TensorShape({channels}));
+    // Output result = ops::Sub(scope.WithOpName("bn_mul"), input, bn_scale);
+    // result = ops::Multiply(scope.WithOpName("bn_shift"), result, bn_offset);
+    auto result =
+        ops::FusedBatchNormV3(
+            scope.WithOpName("bn"),
+            AddQDQV2(scope.WithOpName("qdq_input"), input), bn_scale, bn_offset,
+            bn_mean, bn_var,
+            ops::FusedBatchNormV3::Attrs().IsTraining(false).DataFormat(
+                data_format))
+            .y;
+    return result;
```
Let me check on this. We shouldn't need the QDQ here for TRT >= 8.0

christopherbate replied:
Confirmed we don't need this for TRT >= 8.
I believe this is required for TRT 7.1, confirming now, but in the meantime I've changed logic to require QDQ here for TRT7.1

christopherbate replied:
This is confirmed.

bixia1 commented:
Please explicitly state when this PR is ready for review again. I did a quick check, there are unaddressed comments, and we still use the option use_explicit_precision.

christopherbate commented on tensorflow/compiler/tf2tensorrt/convert/ops/layer_utils.h:
```diff
@@ -366,7 +398,136 @@ class TRTNetworkBuilder {
+    return FindProducerOf(layer->getInput(input_idx));
+  }
+
+  StatusOr<nvinfer1::ILayer*> FindProducerOf(const nvinfer1::ITensor* tensor) {
```
I rearranged it. It does compile fines. Here's an example where you can see all versions of clang and gcc support this [https://godbolt.org/z/sGsbzbqxj](https://godbolt.org/z/sGsbzbqxj), although I can't point to what part of the C++ standard explicitly specifies why this is allowed.

christopherbate commented on tensorflow/compiler/tf2tensorrt/convert/ops/layer_utils.h:
```diff
@@ -366,7 +398,136 @@ class TRTNetworkBuilder {
+    return quant_layer;
+#else
+    StatusOr<nvinfer1::IScaleLayer*> result =
+        this->AddUniformScale(input, scaling_factor, name);
```
No, because the TRT IQuantize layer is taking the reciprocal. I will double check, however.

christopherbate replied:
I confirmed and added some clarifying comments just now.

christopherbate commented on tensorflow/compiler/tf2tensorrt/convert/ops/quantization_ops_test.cc:
```diff
@@ -0,0 +1,542 @@
+};
+}  // namespace
+
+// Test fixture for running graph conversion and evaluating numerical results.
```
I changed it to "This class is a test fixture for..."

christopherbate commented:
> Please explicitly state when this PR is ready for review again. I did a quick check, there are unaddressed comments, and we still use the option use_explicit_precision.

I'm going through it now and marking comments resolved as I make corrections, so that I can keep track. I will post here when I have pushed the actual changes. You can see my last push was 4 days ago.

christopherbate commented on tensorflow/compiler/tf2tensorrt/convert/ops/quantization_ops_test.cc:
```diff
@@ -0,0 +1,542 @@
+
+std::vector<QDQTestOptions> EnumerateQDQTestOptions() {
+  std::vector<QDQTestOptions> result;
+  for (const auto* data_format : {"NCHW", "NHWC"}) {
+    for (auto use_bias : {true, false}) {
+      for (auto qdq_on_output : {false, true}) {
```
How can I add the logic  in the inner most loop?

christopherbate replied:
See comment in main thread, I am using the GTest macro at the bottom, and I believe I need this logic.

christopherbate committed:
[TF:TRT] Enable TRT explicit precision support

christopherbate commented:
I pushed an update to:
- Removed changes from `trt_convert.py`
- Address comments not marked "resolved"

Need to check whether INSTANTIATE_TEST_CASE_P will allow me to create the correct parameters with the logic present in the function I currently have to enumerate test case parameters.

christopherbate commented:
@bixia1 I am using "INSTATITATE_TEST_CASE_P"... that was deprecated and renamed to "INSTATIATE_TEST_SUITE_P". I don't think I can avoid writing a function to enumerate the test suite parameters, since there is some logic that needs to decide exactly how to combine them. It's not an exact Cartesian product of Param1Range x Param2Range x ... x ParamNRange.

christopherbate commented on tensorflow/compiler/tf2tensorrt/convert/ops/quantization_ops_test.cc:
```diff
@@ -0,0 +1,542 @@
+
+class TestQDQSuite : public QDQExplicitTest,
+                     public ::testing::WithParamInterface<QDQTestOptions> {
+  void SetUp() override { VLOG(0) << "Params: " << GetParam(); }
```
I am using this macro, or the updated equivalent. The test name itself doesn't have this information, it only has something like `TestQDQSuiteInst/TestQDQSuite.TestConvMaxpoolConv/5`. Even in our other test fixtures which you linked to in your other comment, the parameters are explicitly printed out. 

However, if there is an error, the test parameters get printed using the stream operator `<<` or `DebugString` method if available, so I removed the log line here.

christopherbate commented:
Updated PR description. All comments are addressed.

bixia1 commented on tensorflow/compiler/tf2tensorrt/convert/trt_optimization_pass.cc:
```diff
@@ -241,6 +243,9 @@ Status TRTOptimizationPass::Init(
     TF_RETURN_IF_ERROR(ProfileStrategyFromName(
         params.at("profile_strategy").s(), &profile_strategy_));
   }
+  if (params.count("use_explicit_precision")) {
+    use_explicit_precision_ = params.at("use_explicit_precision").b();
+  }
```
You removed the change in the converter API for this. But then how do users pass this parameter to the optimization pass? I was under the impression that we don't need this parameter, instead, the TF-TRT internal will figure out whether we will use_explicit_precision based on the version of the TensorRT. Please correct me if I am not right here.

christopherbate replied:
We still need to turn off the constant folding for QDQ ops in grappler

bixia1 replied:
My question was: how would you enable or disable use_explicit_precision? Can you describe?

christopherbate replied:
Okay I thought we were going to do the enable/disable in follow on PR. But I can update this PR

bixia1 replied:
We can do the enable/disable in another PR, but the question is how you are going to enable/disable? 
I don't see why we need to add "use_explicit_precision" to the pass parameter, with the way I have in mind to enable/disable. That is the main reason I am trying to find out how you will enable/disable. Can you describe?

christopherbate replied:
Here's the rough call stack for conversion process.

```
  0 API sets up grappler RewriterConfig (either in Python or on new C++ API side)
  1 Run the TRTOptimizationPass::Optimize.
  2 Calls ConvertAfterShapes (segment and outline)
  3     Creates TrtNodeValidator    
  4     Calls segment::SegmentGraph(uses validator)
  5     For each segment:
  6         Calls GetEngineInfo
  7             Calls ConvertSegmentToGraphDef (outlining)
  8     Calls CreateTRTNode
  9         (Updates graph to use TRTEngineOp)
  10         If STATIC: ConvertGraphDefToEngine
  11
  12 TRTEngineOp::ExecuteAsync
  13    ::GetEngine
  14       if(static_engine)
  15          deserialize_engine
  16       if(!static_engine)
  17          ::BuildEngine
  18             convert::ConvertGraphDeftoEngine
  19    ::ExecuteTrtEngine
```

1. At line 0: We need to know explicitly whether the use because we need to turn on the grappler flag to prevent folding emulation for QDQ. In both Python and new C++ API sides, this can be determined by inspecting user's graph and turning on the flag when QDQ is detected and TRT version is >= 8.0.
2. At lines 3, 10, and 18 we need to know to create Validator/Converter respectively.

Option 1:

In order to minimize the number of places where we "detect QDQ" it seems to make sense that we should just add it as a pass parameter that gets set automatically in the C++/Python API functions rather than in any of lines 3-18. The pass parameter would forward this parameter to ConvertAfterShapes, and finally it would be incorporated into the TRTEngineOp, where it is accessible to lines 10 and 18.

Option 2:

We duplicate the "detect QDQ" logic so that Python/C++ API sets the Grappler option `experimental_disable_folding_quantization_emulation` after detecting QDQ, and in addition the function ConvertAfterShapes also does this sort of detection, so that it can pass on the parameter to TRTEngineOp for lines 10 and 18.


I prefer Option 1 since if there is any problem, it is easier to insert a switch to turn off the new style QDQ conversion and use the old style conversion.

What do you think?

bixia1 replied:
What would be the problem if we always turning on the grappler flag to prevent folding emulation for QDQ?

christopherbate replied:
If you think that's OK, I can do that and infer the parameter at line 3

christopherbate replied:
All done. Let me know when you want me  to squash.

christopherbate commented:
rebased

christopherbate commented:
Pushed changes to add API changes in C++/Python, updated unit tests to use TrtConvert C++ API functions

christopherbate commented:
still working on fixing some tests after converting to the fixture

christopherbate commented:
All done. Let me know when you want me to squash.

bixia1 commented on tensorflow/compiler/tf2tensorrt/convert/ops/quantization_ops_test.cc:
```diff
@@ -0,0 +1,620 @@
+        n_trt_ops++;
+        const auto& attr = node.attr();
+        TRT_ENSURE(attr.at("static_engine").b());
+        VLOG(0) << "Found serialized segment with size "
```
Why do we want to use VLOG(0) here?

christopherbate replied:
fixed

bixia1 commented on tensorflow/compiler/tf2tensorrt/convert/trt_optimization_pass.cc:
```diff
@@ -367,12 +378,27 @@ Status TRTOptimizationPass::Optimize(grappler::Cluster* cluster,
     use_calibration_ = false;
   }
 
+  const bool use_explicit_precision = GraphDefHasQDQNodes(item.graph);
```
I have two questions regarding this:
(1) we don't consider the TensorRT version to enable use_explicit_precision, this looks wrong to me.
For example, a user has a model that uses QuantizeAndDequantizeV2 + TensorRT 7. It was working but now with this PR, he sees an error produced by TensorRT, similar to that in quantization_ops_test.cc  line 435 & 440. Am I right here?
(2)  A user has a model that uses all the OPs in kQuantizationOpNames  + TensorRT8. It was working fine. But now with your change, we only convert one out of the four OPs in kQuantizationOpNames . Will this be a problem?

christopherbate replied:
OK yes, I let me disable this for TRT < 8.0. We can use TRT 7 for very limited use cases, but likely EP mode won't be useful in the context of TF-TRT with TRT7 at this time.

christopherbate replied:
fixed

christopherbate commented:
Completely disabled explicit QDQ tests and grappler options (disable_emulate_quantization_folding) for TRT < 8.0.

christopherbate commented:
@bixia1 Good to squash?

gbaned commented:
@christopherbate Can you please resolve conflicts? Thanks!

christopherbate commented:
> @christopherbate Can you please resolve conflicts? Thanks!

rebased, thanks

bixia1 commented:
@christopherbate please squash and I will approve.

christopherbate commented:
squashed
